### DISCUSSION TITLE
BUG: TextBlock does not center correctly when text is surrounded by equal number of spaces

### DISCUSSION THREAD

**Discussion by kuiperzone** [AvaloniaUI/Avalonia#17205] [General]
NB. Yes I know I should submit it as a bug, but the Github Issue page is not working. The SUBMIT is grayed out.

***
TextBlock does not seem to be center correctly when text is surround by equal number of spaces both sides, including NBSP characters. I suspect it may be trimming one side.

The screenshot below is generated with the following text:

`"       H-Center       \nV-Middle\nLarger", "font-size:larger"`

May seem trivial, but I'm using NBSP characters and nothing would align. Took me two days to figure out the reason.

![image](https://github.com/user-attachments/assets/cddc064c-3d1b-4dc9-ad38-23e090b0ac3d)

Perhaps trimming text is intentional behaviour? I'm guessing it is related to a workbreak algorithm? But I think that text should not be trimmed by default after an explicit line break.

**Comment by Gillibald** [comment]
![Screenshot 2024-10-07 074641](https://github.com/user-attachments/assets/7198d1dd-5ad9-451c-b8c4-40bdab6dce4c)

WPF is producing the same which makes sense. Only trailing spaces are ignored.

**Reply by timunie** [reply]
I don't know if it helps in this situation but you can set `xml:space="preserve"` to disable trimming completely. Alternatively you may want to create a converter that trims out spaces as needed in your case.

**Reply by kuiperzone** [reply]
Then it's not a bug. Thank you for your reply.

@timunie Will look into it. Thank you.

**Reply by kuiperzone** [reply]
I'm seeing examples like the following with regard to preserving spaces:

`<TextBox xml:space="preserve" Name="StructTextBox" Width="129" Height="50"  VerticalAlignment="Top" HorizontalAlignment="Left"     IsHitTestVisible="False" TextWrapping="Wrap" >Fire control Panel      Hello World</TextBox>`

This looks like it relates to parsing the XAML. However, there is no XAML in my project whatsoever. Everything is runtime. In fact, I'm populating TextBlock entirely with the Inline `Run` class and never setting the `Text` property. There's WPF post I saw suggesting `Run` is a possible solution, but screenshot above was created with Run.

Anyway to preserve spaces at runtime?

**Reply by timunie** [reply]
@kuiperzone can you provide a very minimal sample here? I can't really understand how you produce the text.

**Reply by kuiperzone** [reply]
Sure thing:

    protected override void OnOpened(EventArgs e)
    {
        base.OnOpened(e);

        var text = new TextBlock();
        text.HorizontalAlignment = Avalonia.Layout.HorizontalAlignment.Center;
        text.Inlines.Add( new Run("\u00A0\u00A0\u00A0\u00A0H-Center\u00A0\u00A0\u00A0\u00A0\nV-Middle\nLarger") );
        Content = text;
    }


![image](https://github.com/user-attachments/assets/ff43198d-2395-490a-b9bf-ebc56b87255f)

**Reply by kuiperzone** [reply]
NB. Equal spaces on either side of H-Center

Am using NBSP, but same output with normal space. (In my context, the controls are being constructed at runtime, so it's not just a case of changing Text in XAML for me).

**Reply by kuiperzone** [reply]
And same with this:

        var text = new TextBlock();
        text.HorizontalAlignment = Avalonia.Layout.HorizontalAlignment.Center;
        text.Inlines.Add( new Run("\u00A0\u00A0\u00A0\u00A0H-Center\u00A0\u00A0\u00A0\u00A0") );
        text.Inlines.Add( new LineBreak() );
        text.Inlines.Add( new Run("V-Middle") );
        text.Inlines.Add( new LineBreak() );
        text.Inlines.Add( new Run("Larger") );
        Content = text;
